`TestBinomialGenerator` fails due to data race

I'll work on this:
https://travis-ci.org/AkihiroSuda/go_rng/builds/107577141

```
=====Testing for BinomialGenerator begin=====
==================
WARNING: DATA RACE
Read by goroutine 23:
  math/rand.(*rngSource).Int63()
      /tmp/workdir/go/src/math/rand/rng.go:233 +0x32
  math/rand.(*Rand).Int63()
      /tmp/workdir/go/src/math/rand/rand.go:46 +0x54
  math/rand.(*Rand).Float64()
      /tmp/workdir/go/src/math/rand/rand.go:126 +0x31
  github.com/AkihiroSuda/go_rng.UniformGenerator.Float64()
      /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/uniform.go:84 +0x31
  github.com/AkihiroSuda/go_rng.BinomialGenerator.binomial()
      /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/binomial.go:71 +0x36f
  github.com/AkihiroSuda/go_rng.BinomialGenerator.Binomial.func1()
      /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/binomial.go:40 +0x43
Previous write by goroutine 21:
  math/rand.(*rngSource).Int63()
      /tmp/workdir/go/src/math/rand/rng.go:233 +0x48
  math/rand.(*Rand).Int63()
      /tmp/workdir/go/src/math/rand/rand.go:46 +0x54
  math/rand.(*Rand).Float64()
      /tmp/workdir/go/src/math/rand/rand.go:126 +0x31
  github.com/AkihiroSuda/go_rng.UniformGenerator.Float64()
      /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/uniform.go:84 +0x31
  github.com/AkihiroSuda/go_rng.BinomialGenerator.binomial()
      /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/binomial.go:71 +0x36f
  github.com/AkihiroSuda/go_rng.BinomialGenerator.Binomial.func1()
      /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/binomial.go:40 +0x43
Goroutine 23 (running) created at:
  github.com/AkihiroSuda/go_rng.BinomialGenerator.Binomial()
      /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/binomial.go:42 +0x42f
  github.com/AkihiroSuda/go_rng.TestBinomialGenerator()
      /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/rng_test.go:340 +0x20c
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc
Goroutine 21 (running) created at:
  github.com/AkihiroSuda/go_rng.BinomialGenerator.Binomial()
      /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/binomial.go:42 +0x42f
  github.com/AkihiroSuda/go_rng.TestBinomialGenerator()
      /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/rng_test.go:340 +0x20c
  testing.tRunner()
      /tmp/workdir/go/src/testing/testing.go:456 +0xdc
==================
...
==================
panic: runtime error: index out of range
goroutine 1005 [running]:
math/rand.(*rngSource).Int63(0xc8200f5500, 0x6c11abd91d3bae68)
    /tmp/workdir/go/src/math/rand/rng.go:243 +0x27a
math/rand.(*Rand).Int63(0xc8200f0dc0, 0x45c394)
    /tmp/workdir/go/src/math/rand/rand.go:46 +0x55
math/rand.(*Rand).Float64(0xc8200f0dc0, 0x3feb046af6474eec)
    /tmp/workdir/go/src/math/rand/rand.go:126 +0x32
github.com/AkihiroSuda/go_rng.UniformGenerator.Float64(0xc8200f0dc0, 0x8)
    /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/uniform.go:84 +0x32
github.com/AkihiroSuda/go_rng.BinomialGenerator.binomial(0xc8200ae138, 0x3e8, 0x3f947ae147ae147b, 0x0)
    /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/binomial.go:71 +0x370
github.com/AkihiroSuda/go_rng.BinomialGenerator.Binomial.func1(0xc8200ae138, 0x3f947ae147ae147b, 0xc8200a2540)
    /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/binomial.go:40 +0x44
created by github.com/AkihiroSuda/go_rng.BinomialGenerator.Binomial
    /home/travis/gopath/src/github.com/AkihiroSuda/go_rng/binomial.go:42 +0x430
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,0 +1,7 @@
+language: go
+
+go:
+  - 1.5
+
+gobuild_args: -v -race
+

--- a/.travis.yml
+++ b/.travis.yml
@@ -3,5 +3,7 @@ language: go
 go:
   - 1.5
 
-gobuild_args: -v -race
+before_script:
+  - go vet ./...
 
+gobuild_args: -v -race

--- a/README.md
+++ b/README.md
@@ -1,6 +1,8 @@
 go_rng --- 伪随机数生成器库的Go语言实现
 =======================================
 
+[![Build Status](https://travis-ci.org/leesper/go_rng.svg?branch=master)](https://travis-ci.org/leesper/go_rng)
+
 ### A pseudo-random number generator written in Golang v1.1
 
 ### Author: 

--- a/bernoulli.go
+++ b/bernoulli.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -12,7 +12,7 @@ import (
 // UniformGenerator is a random number generator for uniform distribution.
 // The zero value is invalid, use NewBernoulliGenerator to create a generator
 type BernoulliGenerator struct {
-	uniform		*UniformGenerator
+	uniform *UniformGenerator
 }
 
 // NewBernoulliGenerator returns a bernoulli-distribution generator
@@ -20,7 +20,7 @@ type BernoulliGenerator struct {
 // beng := rng.NewBernoulliGenerator(time.Now().UnixNano())
 func NewBernoulliGenerator(seed int64) *BernoulliGenerator {
 	urng := NewUniformGenerator(seed)
-	return &BernoulliGenerator{ urng }
+	return &BernoulliGenerator{urng}
 }
 
 // Bernoulli returns a bool, which is true with probablity 0.5

--- a/beta.go
+++ b/beta.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -11,8 +11,8 @@ import (
 // BetaGenerator is a random number generator for beta distribution.
 // The zero value is invalid, use NewBetaGenerator to create a generator
 type BetaGenerator struct {
-	gamma1	*GammaGenerator
-	gamma2	*GammaGenerator
+	gamma1 *GammaGenerator
+	gamma2 *GammaGenerator
 }
 
 // NewBetaGenerator returns a beta distribution generator
@@ -20,10 +20,10 @@ type BetaGenerator struct {
 // brng := rng.NewBetaGenerator(time.Now().UnixNano())
 func NewBetaGenerator(seed int64) *BetaGenerator {
 	grng1 := NewGammaGenerator(seed)
-	// according to "Numerical Recipes", all distinct 
+	// according to "Numerical Recipes", all distinct
 	//gamma variates must have different random seeds
-	grng2 := NewGammaGenerator(2 * seed + 1)
-	return &BetaGenerator{ grng1, grng2 }
+	grng2 := NewGammaGenerator(2*seed + 1)
+	return &BetaGenerator{grng1, grng2}
 }
 
 // Beta returns a random number of beta distribution (alpha > 0.0 and beta > 0.0)
@@ -40,5 +40,5 @@ func (brng BetaGenerator) Beta(alpha, beta float64) float64 {
 func (brng BetaGenerator) beta(alpha, beta float64) float64 {
 	X := brng.gamma1.Gamma(alpha, 1)
 	Y := brng.gamma2.Gamma(beta, 1)
-	return X / (X + Y) 
+	return X / (X + Y)
 }

--- a/binomial.go
+++ b/binomial.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -12,7 +12,7 @@ import (
 // BinomialGenerator is a random number generator for binomial distribution.
 // The zero value is invalid, use NewBinomialGenerator to create a generator
 type BinomialGenerator struct {
-	uniform		*UniformGenerator
+	uniform *UniformGenerator
 }
 
 // NewBinomialGenerator returns a binomial-distribution generator
@@ -20,7 +20,7 @@ type BinomialGenerator struct {
 // bing := rng.NewBinomialGenerator(time.Now().UnixNano())
 func NewBinomialGenerator(seed int64) *BinomialGenerator {
 	urng := NewUniformGenerator(seed)
-	return &BinomialGenerator{ urng }
+	return &BinomialGenerator{urng}
 }
 
 // Binomial returns a random number X ~ binomial(n, p)
@@ -31,7 +31,7 @@ func (bing BinomialGenerator) Binomial(n int64, p float64) int64 {
 	if n <= 0 {
 		panic(fmt.Sprintf("Invalid parameter n: ", n))
 	}
-	
+
 	workers := 0
 	resChan := make(chan int64)
 	for n > 0 {
@@ -51,10 +51,10 @@ func (bing BinomialGenerator) Binomial(n int64, p float64) int64 {
 			break
 		}
 	}
-	
+
 	var result int64
 	for i := 0; i < workers; i++ {
-		result += <-resChan	
+		result += <-resChan
 	}
 	return result
 }

--- a/binomial.go
+++ b/binomial.go
@@ -26,10 +26,10 @@ func NewBinomialGenerator(seed int64) *BinomialGenerator {
 // Binomial returns a random number X ~ binomial(n, p)
 func (bing BinomialGenerator) Binomial(n int64, p float64) int64 {
 	if !(0.0 <= p && p <= 1.0) {
-		panic(fmt.Sprintf("Invalid probability p: ", p))
+		panic(fmt.Sprintf("Invalid probability p: %f", p))
 	}
 	if n <= 0 {
-		panic(fmt.Sprintf("Invalid parameter n: ", n))
+		panic(fmt.Sprintf("Invalid parameter n: %d", n))
 	}
 
 	workers := 0

--- a/binomial.go
+++ b/binomial.go
@@ -32,31 +32,25 @@ func (bing BinomialGenerator) Binomial(n int64, p float64) int64 {
 		panic(fmt.Sprintf("Invalid parameter n: %d", n))
 	}
 
-	workers := 0
-	resChan := make(chan int64)
-	for n > 0 {
-		if n > 1000 {
+	if n > 1000 {
+		workers := 0
+		resChan := make(chan int64)
+		for n > 0 {
 			go func() {
 				res := bing.binomial(1000, p)
 				resChan <- res
 			}()
 			n -= 1000
 			workers++
-		} else {
-			go func() {
-				res := bing.binomial(n, p)
-				resChan <- res
-			}()
-			workers++
-			break
 		}
+		var result int64
+		for i := 0; i < workers; i++ {
+			result += <-resChan
+		}
+		return result
+	} else {
+		return bing.binomial(n, p)
 	}
-
-	var result int64
-	for i := 0; i < workers; i++ {
-		result += <-resChan
-	}
-	return result
 }
 
 func (bing BinomialGenerator) binomial(n int64, p float64) int64 {

--- a/cauchy.go
+++ b/cauchy.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -13,17 +13,16 @@ import (
 // CauchyGenerator is a random number generator for cauchy distribution.
 // The zero value is invalid, use NewCauchyGenerator to create a generator
 type CauchyGenerator struct {
-	uniform		*UniformGenerator
-}	
+	uniform *UniformGenerator
+}
 
 // NewCauchyGenerator returns a cauchy-distribution generator
 // it is recommended using time.Now().UnixNano() as the seed, for example:
 // crng := rng.NewCauchyGenerator(time.Now().UnixNano())
 func NewCauchyGenerator(seed int64) *CauchyGenerator {
 	urng := NewUniformGenerator(seed)
-	return &CauchyGenerator{ urng }
+	return &CauchyGenerator{urng}
 }
-
 
 // Cauchy returns a random number of cauchy distribution
 func (crng CauchyGenerator) Cauchy(x0, gamma float64) float64 {
@@ -39,5 +38,5 @@ func (crng CauchyGenerator) StandardCauchy() float64 {
 }
 
 func (crng CauchyGenerator) cauchy(x0, gamma float64) float64 {
-	return x0 + gamma * math.Tan(math.Pi * (crng.uniform.Float64() - 0.5))
+	return x0 + gamma*math.Tan(math.Pi*(crng.uniform.Float64()-0.5))
 }

--- a/cauchy.go
+++ b/cauchy.go
@@ -27,7 +27,7 @@ func NewCauchyGenerator(seed int64) *CauchyGenerator {
 // Cauchy returns a random number of cauchy distribution
 func (crng CauchyGenerator) Cauchy(x0, gamma float64) float64 {
 	if !(gamma > 0.0) {
-		panic(fmt.Sprintf("Invalid parameter gamma: .2f", gamma))
+		panic(fmt.Sprintf("Invalid parameter gamma: %.2f", gamma))
 	}
 	return crng.cauchy(x0, gamma)
 }

--- a/chi_square.go
+++ b/chi_square.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -11,7 +11,7 @@ import (
 // ChiSquaredGenerator is a random number generator for chi-squared distribution.
 // The zero value is invalid, use NewChiSquaredGenerator to create a generator
 type ChiSquaredGenerator struct {
-	gamma	*GammaGenerator
+	gamma *GammaGenerator
 }
 
 // NewChiSquaredGenerator returns a chi-squared distribution generator
@@ -19,7 +19,7 @@ type ChiSquaredGenerator struct {
 // crng := rng.NewChiSquaredGenerator(time.Now().UnixNano())
 func NewChiSquaredGenerator(seed int64) *ChiSquaredGenerator {
 	grng := NewGammaGenerator(seed)
-	return &ChiSquaredGenerator{ grng }
+	return &ChiSquaredGenerator{grng}
 }
 
 // ChiSquared returns a random number of chi-squared distribution (freedom > 0)
@@ -32,5 +32,5 @@ func (crng ChiSquaredGenerator) ChiSquared(freedom int64) float64 {
 
 func (crng ChiSquaredGenerator) chisquared(freedom int64) float64 {
 	real_val := float64(freedom)
-	return crng.gamma.Gamma(real_val / 2.0, 2.0)
+	return crng.gamma.Gamma(real_val/2.0, 2.0)
 }

--- a/exponential.go
+++ b/exponential.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -13,16 +13,15 @@ import (
 // ExpGenerator is a random number generator for exponential distribution.
 // The zero value is invalid, use NewExpGenerator to create a generator
 type ExpGenerator struct {
-	uniform		*UniformGenerator
+	uniform *UniformGenerator
 }
-
 
 // NewExpGenerator returns a exponential-distribution generator
 // it is recommended using time.Now().UnixNano() as the seed, for example:
 // erng := rng.NewExpGenerator(time.Now().UnixNano())
 func NewExpGenerator(seed int64) *ExpGenerator {
 	urng := NewUniformGenerator(seed)
-	return &ExpGenerator{ urng }
+	return &ExpGenerator{urng}
 }
 
 // Exp returns a random number of exponential distribution
@@ -34,5 +33,5 @@ func (erng ExpGenerator) Exp(lambda float64) float64 {
 }
 
 func (erng ExpGenerator) exp(lambda float64) float64 {
-	return -math.Log(1 - erng.uniform.Float64()) / lambda
+	return -math.Log(1-erng.uniform.Float64()) / lambda
 }

--- a/fisher_f.go
+++ b/fisher_f.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -11,7 +11,7 @@ import (
 // FisherFGenerator is a random number generator for Fisher's F distribution.
 // The zero value is invalid, use NewFisherFGenerator to create a generator
 type FisherFGenerator struct {
-	beta	*BetaGenerator
+	beta *BetaGenerator
 }
 
 // NewFisherFGenerator returns a Fisher's F distribution generator
@@ -19,7 +19,7 @@ type FisherFGenerator struct {
 // frng := rng.NewFisherFGenerator(time.Now().UnixNano())
 func NewFisherFGenerator(seed int64) *FisherFGenerator {
 	brng := NewBetaGenerator(seed)
-	return &FisherFGenerator{ brng }
+	return &FisherFGenerator{brng}
 }
 
 // Fisher returns a random number of Fisher's F distribution (d1 > 0 and d2 > 0)
@@ -36,6 +36,6 @@ func (frng FisherFGenerator) Fisher(d1, d2 int64) float64 {
 func (frng FisherFGenerator) fisher(d1, d2 int64) float64 {
 	f1 := float64(d1)
 	f2 := float64(d2)
-	X := frng.beta.Beta(f1 / 2.0, f2 / 2.0)
-	return f2 * X / ( f1 * (1 - X) )
+	X := frng.beta.Beta(f1/2.0, f2/2.0)
+	return f2 * X / (f1 * (1 - X))
 }

--- a/gamma.go
+++ b/gamma.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -12,7 +12,7 @@ import (
 // GammaGenerator is a random number generator for gamma distribution.
 // The zero value is invalid, use NewGammaGenerator to create a generator
 type GammaGenerator struct {
-	uniform 	*UniformGenerator
+	uniform *UniformGenerator
 }
 
 // NewGammaGenerator returns a gamma distribution generator
@@ -20,15 +20,15 @@ type GammaGenerator struct {
 // grng := rng.NewGammaGenerator(time.Now().UnixNano())
 func NewGammaGenerator(seed int64) *GammaGenerator {
 	urng := NewUniformGenerator(seed)
-	return &GammaGenerator{ urng }
+	return &GammaGenerator{urng}
 }
 
 // Gamma returns a random number of gamma distribution (alpha > 0.0 and beta > 0.0)
 func (grng GammaGenerator) Gamma(alpha, beta float64) float64 {
-	if !(alpha > 0.0) || !(beta > 0.0){
+	if !(alpha > 0.0) || !(beta > 0.0) {
 		panic(fmt.Sprintf("Invalid parameter alpha %.2f beta %.2f", alpha, beta))
 	}
-	
+
 	return grng.gamma(alpha, beta)
 }
 
@@ -38,22 +38,22 @@ func (grng GammaGenerator) gamma(alpha, beta float64) float64 {
 	if alpha > 1.0 {
 		// Use R.C.H Cheng "The generation of Gamma variables with
 		// non-integral shape parameters", Applied Statistics, (1977), 26, No. 1, p71-74
-		
-		ainv := math.Sqrt(2.0 * alpha - 1.0)
+
+		ainv := math.Sqrt(2.0*alpha - 1.0)
 		bbb := alpha - math.Log(4.0)
 		ccc := alpha + ainv
-		
+
 		for {
 			u1 := grng.uniform.Float64()
 			if !(1e-7 < u1 && u1 < .9999999) {
 				continue
 			}
 			u2 := 1.0 - grng.uniform.Float64()
-			v := math.Log(u1 / (1.0 - u1)) / ainv
+			v := math.Log(u1/(1.0-u1)) / ainv
 			x := alpha * math.Exp(v)
 			z := u1 * u1 * u2
-			r := bbb + ccc * v - x
-			if r + MAGIC_CONST - 4.5 * z >= 0.0 || r >= math.Log(z) {
+			r := bbb + ccc*v - x
+			if r+MAGIC_CONST-4.5*z >= 0.0 || r >= math.Log(z) {
 				return x * beta
 			}
 		}
@@ -71,13 +71,13 @@ func (grng GammaGenerator) gamma(alpha, beta float64) float64 {
 			b := (math.E + alpha) / math.E
 			p := b * u
 			if p <= 1.0 {
-				x = math.Pow(p, 1.0 / alpha)
+				x = math.Pow(p, 1.0/alpha)
 			} else {
 				x = -math.Log((b - p) / alpha)
 			}
 			u1 := grng.uniform.Float64()
 			if p > 1.0 {
-				if u1 <= math.Pow(x, alpha - 1.0) {
+				if u1 <= math.Pow(x, alpha-1.0) {
 					break
 				}
 			} else if u1 <= math.Exp(-x) {

--- a/gamma.go
+++ b/gamma.go
@@ -86,5 +86,4 @@ func (grng GammaGenerator) gamma(alpha, beta float64) float64 {
 		}
 		return x * beta
 	}
-	panic("This place cannot be reached")
 }

--- a/gauss.go
+++ b/gauss.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -11,7 +11,7 @@ import (
 // GaussianGenerator is a random number generator for gaussian distribution.
 // The zero value is invalid, use NewGaussianGenerator to create a generator
 type GaussianGenerator struct {
-	uniform		*UniformGenerator
+	uniform *UniformGenerator
 }
 
 // NewGaussianGenerator returns a gaussian-distribution generator
@@ -19,7 +19,7 @@ type GaussianGenerator struct {
 // crng := rng.NewGaussianGenerator(time.Now().UnixNano())
 func NewGaussianGenerator(seed int64) *GaussianGenerator {
 	urng := NewUniformGenerator(seed)
-	return &GaussianGenerator{ urng }
+	return &GaussianGenerator{urng}
 }
 
 // StdGaussian returns a random number of standard gaussian distribution
@@ -29,7 +29,7 @@ func (grng GaussianGenerator) StdGaussian() float64 {
 
 // Gaussian returns a random number of gaussian distribution Gauss(mean, stddev^2)
 func (grng GaussianGenerator) Gaussian(mean, stddev float64) float64 {
-	return mean + stddev * grng.gaussian()
+	return mean + stddev*grng.gaussian()
 }
 
 func (grng GaussianGenerator) gaussian() float64 {
@@ -38,8 +38,7 @@ func (grng GaussianGenerator) gaussian() float64 {
 	for r >= 1 || r == 0 {
 		x = grng.uniform.Float64Range(-1.0, 1.0)
 		y = grng.uniform.Float64Range(-1.0, 1.0)
-		r = x * x + y * y
+		r = x*x + y*y
 	}
-	return x * math.Sqrt(-2 * math.Log(r) / r)
+	return x * math.Sqrt(-2*math.Log(r)/r)
 }
-

--- a/geometry.go
+++ b/geometry.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -13,7 +13,7 @@ import (
 // GeometricGenerator is a random number generator for geometric distribution.
 // The zero value is invalid, use NewGeometryGenerator to create a generator
 type GeometricGenerator struct {
-	uniform		*UniformGenerator
+	uniform *UniformGenerator
 }
 
 // NewGeometricGenerator returns a geometric-distribution generator
@@ -21,18 +21,18 @@ type GeometricGenerator struct {
 // grng := rng.NewGeometricGenerator(time.Now().UnixNano())
 func NewGeometricGenerator(seed int64) *GeometricGenerator {
 	urng := NewUniformGenerator(seed)
-	return &GeometricGenerator{ urng }
+	return &GeometricGenerator{urng}
 }
 
 // Geometric returns a random number X ~ binomial(n, p)
 func (grng GeometricGenerator) Geometric(p float64) int64 {
 	if !(0.0 <= p && p <= 1.0) {
-		panic(fmt.Sprintf("Invalid probability p: %.2f", p))	
+		panic(fmt.Sprintf("Invalid probability p: %.2f", p))
 	}
 	return grng.geometric(p)
 }
 
 func (grng GeometricGenerator) geometric(p float64) int64 {
 	// ceil( ln(U)/ln(1-p) ), where U ~ Uniform(0, 1)
-	return int64(math.Floor( math.Log(grng.uniform.Float64()) / math.Log(1.0 - p)) )
+	return int64(math.Floor(math.Log(grng.uniform.Float64()) / math.Log(1.0-p)))
 }

--- a/logistic.go
+++ b/logistic.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -12,8 +12,8 @@ import (
 
 // LogisticGenerator is a random number generator for logistic distribution.
 // The zero value is invalid, use NewLogisticGenerator to create a generator
-type LogisticGenerator struct {	
-	uniform		*UniformGenerator
+type LogisticGenerator struct {
+	uniform *UniformGenerator
 }
 
 // NewLogisticGenerator returns a logistic-distribution generator
@@ -21,17 +21,17 @@ type LogisticGenerator struct {
 // lrng := rng.NewLogisticGenerator(time.Now().UnixNano())
 func NewLogisticGenerator(seed int64) *LogisticGenerator {
 	urng := NewUniformGenerator(seed)
-	return &LogisticGenerator{ urng }
+	return &LogisticGenerator{urng}
 }
 
 // Logistic returns a random number of logistic distribution
 func (lrng LogisticGenerator) Logistic(mu, s float64) float64 {
 	if !(s > 0.0) {
 		panic(fmt.Sprintf("Invalid parameter s: %.2f", s))
-	}	
+	}
 	return lrng.logistic(mu, s)
 }
 
 func (lrng LogisticGenerator) logistic(mu, s float64) float64 {
-	return mu + s * math.Log( lrng.uniform.Float64() / (1 - lrng.uniform.Float64()) ) 
+	return mu + s*math.Log(lrng.uniform.Float64()/(1-lrng.uniform.Float64()))
 }

--- a/lognormal.go
+++ b/lognormal.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -11,7 +11,7 @@ import (
 // LognormalGenerator is a random number generator for lognormal distribution.
 // The zero value is invalid, use NewLognormalGenerator to create a generator
 type LognormalGenerator struct {
-	gauss	*GaussianGenerator
+	gauss *GaussianGenerator
 }
 
 // NewLognormalGenerator returns a lognormal-distribution generator
@@ -19,10 +19,10 @@ type LognormalGenerator struct {
 // crng := rng.NewLognormalGenerator(time.Now().UnixNano())
 func NewLognormalGenerator(seed int64) *LognormalGenerator {
 	grng := NewGaussianGenerator(seed)
-	return &LognormalGenerator{ grng }
+	return &LognormalGenerator{grng}
 }
 
 // Lognormal return a random number of lognormal distribution
 func (lnng LognormalGenerator) Lognormal(mean, stddev float64) float64 {
-	return math.Exp(mean + stddev * lnng.gauss.StdGaussian()) 
+	return math.Exp(mean + stddev*lnng.gauss.StdGaussian())
 }

--- a/pareto.go
+++ b/pareto.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -12,7 +12,7 @@ import (
 // ParetoGenerator is a random number generator for type I pareto distribution.
 // The zero value is invalid, use NewParetoGenerator to create a generator
 type ParetoGenerator struct {
-	uniform 	*UniformGenerator
+	uniform *UniformGenerator
 }
 
 // NewParetoGenerator returns a type I pareto-distribution generator
@@ -20,7 +20,7 @@ type ParetoGenerator struct {
 // crng := rng.NewParetoGenerator(time.Now().UnixNano())
 func NewParetoGenerator(seed int64) *ParetoGenerator {
 	urng := NewUniformGenerator(seed)
-	return &ParetoGenerator{ urng }
+	return &ParetoGenerator{urng}
 }
 
 // Pareto returns a random number of type I pareto distribution (alpha > 0,0)
@@ -28,10 +28,10 @@ func (prng ParetoGenerator) Pareto(alpha float64) float64 {
 	if !(alpha > 0.0) {
 		panic(fmt.Sprintf("Invalid parameter alpha: %.2f", alpha))
 	}
-	
+
 	return prng.pareto(alpha)
 }
 
 func (prng ParetoGenerator) pareto(alpha float64) float64 {
-	return math.Pow(1 - prng.uniform.Float64(), -1.0 / alpha) - 1.0
+	return math.Pow(1-prng.uniform.Float64(), -1.0/alpha) - 1.0
 }

--- a/poisson.go
+++ b/poisson.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -7,14 +7,14 @@ package rng
 
 import (
 	"fmt"
-	"time"
 	"math"
+	"time"
 )
 
 // PoissonGenerator is a random number generator for possion distribution.
 // The zero value is invalid, use NewPoissonGenerator to create a generator
 type PoissonGenerator struct {
-	uniform		*UniformGenerator
+	uniform *UniformGenerator
 }
 
 // NewPoissonGenerator returns a possion-distribution generator
@@ -22,7 +22,7 @@ type PoissonGenerator struct {
 // prng := rng.NewPoissonGenerator(time.Now().UnixNano())
 func NewPoissonGenerator(seed int64) *PoissonGenerator {
 	urng := NewUniformGenerator(time.Now().UnixNano())
-	return &PoissonGenerator{ urng }
+	return &PoissonGenerator{urng}
 }
 
 // Poisson returns a random number of possion distribution
@@ -38,7 +38,7 @@ func (prng PoissonGenerator) poisson(lambda float64) int64 {
 	L := math.Pow(math.E, -lambda)
 	var k int64 = 0
 	var p float64 = 1.0
-	
+
 	for p > L {
 		k++
 		p *= prng.uniform.Float64()

--- a/student_t.go
+++ b/student_t.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -12,8 +12,8 @@ import (
 // StudentTGenerator is a random number generator for student-t distribution.
 // The zero value is invalid, use NewStudentTGenerator to create a generator
 type StudentTGenerator struct {
-	gaussian	*GaussianGenerator
-	chisquared	*ChiSquaredGenerator
+	gaussian   *GaussianGenerator
+	chisquared *ChiSquaredGenerator
 }
 
 // NewStudentTGenerator returns a student-t distribution generator
@@ -22,8 +22,8 @@ type StudentTGenerator struct {
 func NewStudentTGenerator(seed int64) *StudentTGenerator {
 	// make sure they have different random seeds
 	grng := NewGaussianGenerator(seed)
-	crng := NewChiSquaredGenerator(2 * seed + 1)
-	return &StudentTGenerator{ grng, crng }
+	crng := NewChiSquaredGenerator(2*seed + 1)
+	return &StudentTGenerator{grng, crng}
 }
 
 // Student returns a random number of student-t distribution (freedom > 0.0)
@@ -36,6 +36,6 @@ func (stng StudentTGenerator) Student(freedom int64) float64 {
 
 // inspired by http://www.xycoon.com/stt_random.htm
 func (stng StudentTGenerator) student(freedom int64) float64 {
-	X := stng.gaussian.StdGaussian() / math.Sqrt( stng.chisquared.ChiSquared(freedom) / float64(freedom) )
+	X := stng.gaussian.StdGaussian() / math.Sqrt(stng.chisquared.ChiSquared(freedom)/float64(freedom))
 	return X
 }

--- a/uniform.go
+++ b/uniform.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -15,15 +15,15 @@ import (
 // UniformGenerator is a random number generator for uniform distribution.
 // The zero value is invalid, use NewUniformGenerator to create a generator
 type UniformGenerator struct {
-	mu      *sync.Mutex
-	rd	*rand.Rand
+	mu *sync.Mutex
+	rd *rand.Rand
 }
 
 // NewUniformGenerator returns a uniform-distribution generator
 // it is recommended using time.Now().UnixNano() as the seed, for example:
 // urng := rng.NewUniformGenerator(time.Now().UnixNano())
 func NewUniformGenerator(seed int64) *UniformGenerator {
-	return &UniformGenerator{ mu: new(sync.Mutex), rd: rand.New(rand.NewSource(seed)) }	
+	return &UniformGenerator{mu: new(sync.Mutex), rd: rand.New(rand.NewSource(seed))}
 }
 
 // Int32 returns a random uint32
@@ -65,10 +65,10 @@ func (ung UniformGenerator) Int32Range(a, b int32) int32 {
 	if b <= a {
 		panic(fmt.Sprintf("Illegal parameter a, b: %d, %d", a, b))
 	}
-	if b - a > math.MaxInt32 {
+	if b-a > math.MaxInt32 {
 		panic(fmt.Sprintf("Illegal parameter a, b: %d, %d", a, b))
 	}
-	return a + ung.Int32n(b - a)
+	return a + ung.Int32n(b-a)
 }
 
 // Int64Range returns a random uint64 in [a, b)
@@ -76,20 +76,20 @@ func (ung UniformGenerator) Int64Range(a, b int64) int64 {
 	if b <= a {
 		panic(fmt.Sprintf("Illegal parameter a, b: %d, %d", a, b))
 	}
-	if b - a > math.MaxInt32 {
+	if b-a > math.MaxInt32 {
 		panic(fmt.Sprintf("Illegal parameter a, b: %d, %d", a, b))
 	}
-	return a + ung.Int64n(b - a)
+	return a + ung.Int64n(b-a)
 }
 
-// Float32 returns a random float32 in [0.0, 1.0) 
+// Float32 returns a random float32 in [0.0, 1.0)
 func (ung UniformGenerator) Float32() float32 {
 	ung.mu.Lock()
 	defer ung.mu.Unlock()
 	return ung.rd.Float32()
 }
 
-// Float64 returns a random float64 in [0.0, 1.0) 
+// Float64 returns a random float64 in [0.0, 1.0)
 func (ung UniformGenerator) Float64() float64 {
 	ung.mu.Lock()
 	defer ung.mu.Unlock()
@@ -97,19 +97,19 @@ func (ung UniformGenerator) Float64() float64 {
 }
 
 // Float32Range returns a random float32 in [a, b)
-func (ung UniformGenerator) Float32Range(a, b float32) float32 {	
-	if !( a < b ) {
+func (ung UniformGenerator) Float32Range(a, b float32) float32 {
+	if !(a < b) {
 		panic(fmt.Sprintf("Invalid range: %.2f ~ %.2f", a, b))
 	}
-	return a + ung.Float32() * (b - a)
+	return a + ung.Float32()*(b-a)
 }
 
 // Float32Range returns a random float32 in [a, b)
 func (ung UniformGenerator) Float64Range(a, b float64) float64 {
-	if !( a < b ) {
+	if !(a < b) {
 		panic(fmt.Sprintf("Invalid range: %.2f ~ %.2f", a, b))
 	}
-	return a + ung.Float64() * (b - a)
+	return a + ung.Float64()*(b-a)
 }
 
 // Float32n returns a random float32 in [0.0, n)
@@ -126,7 +126,7 @@ func (ung UniformGenerator) Float64n(n float64) float64 {
 func (ung UniformGenerator) Shuffle(arr []interface{}) {
 	N := len(arr)
 	for i := range arr {
-		r := int32(i) + ung.Int32n( int32(N - i) )
+		r := int32(i) + ung.Int32n(int32(N-i))
 		arr[i], arr[r] = arr[r], arr[i]
 	}
 }
@@ -137,7 +137,7 @@ func (ung UniformGenerator) ShuffleRange(arr []interface{}, low, high int) {
 		panic(fmt.Sprintf("Illegal subarray range %d ~ %d", low, high))
 	}
 	for i := low; i <= high; i++ {
-		r := int32(i) + ung.Int32n( int32(high - i + 1) )
+		r := int32(i) + ung.Int32n(int32(high-i+1))
 		arr[i], arr[r] = arr[r], arr[i]
 	}
 }

--- a/uniform.go
+++ b/uniform.go
@@ -9,11 +9,13 @@ import (
 	"fmt"
 	"math"
 	"math/rand"
+	"sync"
 )
 
 // UniformGenerator is a random number generator for uniform distribution.
 // The zero value is invalid, use NewUniformGenerator to create a generator
 type UniformGenerator struct {
+	mu      *sync.Mutex
 	rd	*rand.Rand
 }
 
@@ -21,16 +23,20 @@ type UniformGenerator struct {
 // it is recommended using time.Now().UnixNano() as the seed, for example:
 // urng := rng.NewUniformGenerator(time.Now().UnixNano())
 func NewUniformGenerator(seed int64) *UniformGenerator {
-	return &UniformGenerator{ rand.New(rand.NewSource(seed)) }	
+	return &UniformGenerator{ mu: new(sync.Mutex), rd: rand.New(rand.NewSource(seed)) }	
 }
 
 // Int32 returns a random uint32
 func (ung UniformGenerator) Int32() int32 {
+	ung.mu.Lock()
+	defer ung.mu.Unlock()
 	return ung.rd.Int31()
 }
 
 // Int64 returns a random uint64
 func (ung UniformGenerator) Int64() int64 {
+	ung.mu.Lock()
+	defer ung.mu.Unlock()
 	return ung.rd.Int63()
 }
 
@@ -39,7 +45,8 @@ func (ung UniformGenerator) Int32n(n int32) int32 {
 	if n <= 0 {
 		panic(fmt.Sprintf("Illegal parameter n: %d", n))
 	}
-
+	ung.mu.Lock()
+	defer ung.mu.Unlock()
 	return ung.rd.Int31n(n)
 }
 
@@ -48,7 +55,8 @@ func (ung UniformGenerator) Int64n(n int64) int64 {
 	if n <= 0 {
 		panic(fmt.Sprintf("Illegal parameter n: %d", n))
 	}
-	
+	ung.mu.Lock()
+	defer ung.mu.Unlock()
 	return ung.rd.Int63n(n)
 }
 
@@ -76,11 +84,15 @@ func (ung UniformGenerator) Int64Range(a, b int64) int64 {
 
 // Float32 returns a random float32 in [0.0, 1.0) 
 func (ung UniformGenerator) Float32() float32 {
+	ung.mu.Lock()
+	defer ung.mu.Unlock()
 	return ung.rd.Float32()
 }
 
 // Float64 returns a random float64 in [0.0, 1.0) 
 func (ung UniformGenerator) Float64() float64 {
+	ung.mu.Lock()
+	defer ung.mu.Unlock()
 	return ung.rd.Float64()
 }
 

--- a/weibull.go
+++ b/weibull.go
@@ -1,4 +1,4 @@
-// Package rng implements a series of pseudo-random number generator 
+// Package rng implements a series of pseudo-random number generator
 // based on a variety of common probability distributions
 // Author: Leesper
 // Email: pascal7718@gmail.com 394683518@qq.com
@@ -13,7 +13,7 @@ import (
 // WeibullGenerator is a random number generator for type weibull distribution.
 // The zero value is invalid, use NewWeibullGenerator to create a generator
 type WeibullGenerator struct {
-	uniform		*UniformGenerator
+	uniform *UniformGenerator
 }
 
 // NewWeibullGenerator returns a weibull-distribution generator
@@ -21,22 +21,22 @@ type WeibullGenerator struct {
 // wrng := rng.NewWeibullGenerator(time.Now().UnixNano())
 func NewWeibullGenerator(seed int64) *WeibullGenerator {
 	urng := NewUniformGenerator(seed)
-	return &WeibullGenerator{ urng }
+	return &WeibullGenerator{urng}
 }
 
 // Weibull returns a random number of weibull distribution (lambda > 0.0 and k > 0.0)
 func (wrng WeibullGenerator) Weibull(lambda, k float64) float64 {
 	if !(lambda > 0.0) {
 		panic(fmt.Sprintf("Invalid parameter lambda: %.2f", lambda))
-	}	
-	
+	}
+
 	if !(k > 0.0) {
 		panic(fmt.Sprintf("Invalid parameter k: %.2f", k))
 	}
-	
+
 	return wrng.weibull(lambda, k)
 }
 
 func (wrng WeibullGenerator) weibull(lambda, k float64) float64 {
-	return lambda * math.Pow( -math.Log(1 - wrng.uniform.Float64()), 1/k )
+	return lambda * math.Pow(-math.Log(1-wrng.uniform.Float64()), 1/k)
 }

